File: notebooks/fetch_trisk_results.R
# Define arguments
ARGS <- list(
  BASELINE_SCENARIO = NULL,  # Set to NULL to include all baseline scenarios
  TARGET_SCENARIO = NULL,    # Set to NULL to include all target scenarios
  SCENARIO_GEOGRAPHY = NULL, # Set to NULL to include all geographies
  SHOCK_YEAR = NULL,        # Set to NULL to include all shock years
  CCS_STATUS = NULL,        # Set to NULL to include all CCS statuses
  SCENARIO_PROVIDER = NULL  # Set to NULL to include all providers
)

# Define valid arguments
VALID_ARGS <- names(ARGS)

# Load required libraries
library(mlflow)
library(dplyr)
library(readr)
library(purrr)

# Hardcoded MLflow tracking URI and experiment name
mlflow_tracking_uri <- "https://mlflow.1in1000.com:443"
experiment_name <- "age_impact_v1"

# Initialize MLflow
mlflow_set_tracking_uri(mlflow_tracking_uri)
mlflow_set_experiment(experiment_name)

# Function to check if a run matches the filter parameters
matches_filters <- function(run_data, filters) {
  if (length(filters) == 0) return(TRUE)
  
  # Get run parameters
  params <- run_data$data$params
  if (is.null(params) || nrow(params) == 0) return(FALSE)
  
  # Convert params to named list
  param_list <- setNames(params$value, params$key)
  
  # Check if all filters match
  all(sapply(names(filters), function(param_name) {
    param_value <- filters[[param_name]]
    !is.null(param_list[[param_name]]) && param_list[[param_name]] == param_value
  }))
}

# Function to download and combine artifacts from a run
download_run_artifacts <- function(run_id, output_dir) {
  # Create run-specific directory
  run_dir <- file.path(output_dir, run_id)
  dir.create(run_dir, recursive = TRUE, showWarnings = FALSE)
  
  # Get run info
  run_data <- mlflow_get_run(run_id)
  
  # Download artifacts
  tryCatch({
    # Start a new run context
    mlflow_start_run(run_id = run_id)
    
    message(sprintf("Listing artifacts for run %s...", run_id))
    artifacts <- mlflow_list_artifacts(run_id)
    
    if (nrow(artifacts) == 0) {
      message(sprintf("No artifacts found for run %s", run_id))
      return(FALSE)
    }
    
    message(sprintf("Found %d artifacts for run %s", nrow(artifacts), run_id))
    
    for (i in 1:nrow(artifacts)) {
      artifact_path <- artifacts$path[i]
      message(sprintf("Processing artifact: %s", artifact_path))
      
      if (grepl("\\.csv$", artifact_path)) {
        # Download CSV file
        local_path <- file.path(run_dir, basename(artifact_path))
        message(sprintf("Attempting to download to: %s", normalizePath(local_path)))
        
        tryCatch({
          mlflow_download_artifacts(run_id, artifact_path, local_path)
          message(sprintf("Successfully downloaded artifact to: %s", normalizePath(local_path)))
          
          # Read and add run parameters
          data <- read_csv(local_path)
          params <- run_data$data$params
          if (!is.null(params) && nrow(params) > 0) {
            param_list <- setNames(params$value, params$key)
            for (param_name in names(param_list)) {
              data[[param_name]] <- param_list[[param_name]]
            }
          }
          
          # Save enriched data
          write_csv(data, local_path)
          message(sprintf("Successfully processed and saved data to: %s", normalizePath(local_path)))
        }, error = function(e) {
          message(sprintf("Error processing artifact %s: %s", artifact_path, e$message))
        })
      } else {
        message(sprintf("Skipping non-CSV artifact: %s", artifact_path))
      }
    }
    
    # End the run
    mlflow_end_run()
    return(TRUE)
  }, error = function(e) {
    message(sprintf("Error downloading artifacts for run %s: %s", run_id, e$message))
    # Make sure to end the run even if there's an error
    tryCatch(mlflow_end_run(), error = function(e) {})
    return(FALSE)
  })
}

# Main execution
main <- function() {
  # Create output directory
  output_dir <- "trisk_results"
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  
  # Get all runs from the experiment
  runs <- mlflow_search_runs(experiment_ids = list(mlflow_get_experiment()$experiment_id))
  
  # Convert arguments to filters
  filters <- list()
  
  for (param in VALID_ARGS) {
    if (!is.null(ARGS[[param]])) {
      # Convert parameter name to lowercase for MLflow comparison
      filters[[tolower(param)]] <- ARGS[[param]]
    }
  }
  
  # Filter and process runs
  successful_downloads <- 0
  failed_downloads <- 0
  
  for (i in 1:nrow(runs)) {
    run_id <- runs$run_uuid[i]
    run_data <- mlflow_get_run(run_id)
    
    if (matches_filters(run_data, filters)) {
      message(sprintf("Processing run %s", run_id))
      if (download_run_artifacts(run_id, output_dir)) {
        successful_downloads <- successful_downloads + 1
      } else {
        failed_downloads <- failed_downloads + 1
      }
    }
  }
  
  # Print summary
  message("\nDownload Summary:")
  message(sprintf("Total runs found: %d", nrow(runs)))
  message(sprintf("Runs matching filters: %d", successful_downloads + failed_downloads))
  message(sprintf("Successful downloads: %d", successful_downloads))
  message(sprintf("Failed downloads: %d", failed_downloads))
  message(sprintf("\nResults saved in: %s", normalizePath(output_dir)))
}

# Run the main function
main() 

File: notebooks/fetch_trisk_results.py
import mlflow
import pandas as pd
from typing import List, Dict
import json
import os
from pathlib import Path
import shutil

# MLflow configuration
mlflow_tracking_uri = "https://mlflow.1in1000.com:443"
experiment_name = "age_impact_v1"


def fetch_successful_runs() -> List[Dict]:
    """
    Fetch all successful runs from MLflow and their parameters.
    A run is considered successful if it doesn't have an 'error' parameter.
    """
    # Set tracking URI
    mlflow.set_tracking_uri(mlflow_tracking_uri)

    # Get experiment
    experiment = mlflow.get_experiment_by_name(experiment_name)
    if experiment is None:
        raise ValueError(f"Experiment '{experiment_name}' not found")

    # Search for all runs in the experiment
    runs = mlflow.search_runs(experiment_ids=[experiment.experiment_id])

    # Filter for successful runs (no error parameter)
    successful_runs = []
    for _, run in runs.iterrows():
        # Get run details
        run_data = mlflow.get_run(run.run_id)
        params = run_data.data.params

        # Check if run has an error parameter
        if "error" not in params:
            successful_runs.append(
                {
                    "run_id": run.run_id,
                    "parameters": params,
                    "metrics": run_data.data.metrics,
                    "status": run_data.info.status,
                    "start_time": run_data.info.start_time,
                    "end_time": run_data.info.end_time,
                }
            )

    return successful_runs


def filter_runs_by_target_scenario(
    runs: List[Dict], target_scenario: str
) -> List[Dict]:
    """
    Filter runs by target scenario.
    """
    return [
        run
        for run in runs
        if run["parameters"].get("target_scenario") == target_scenario
    ]


def download_run_artifacts(run_id: str, output_dir: str) -> Dict[str, str]:
    """
    Download artifacts for a specific run and rename them with run ID.
    Returns a dictionary with paths to the downloaded files.
    """
    # Create output directory if it doesn't exist
    os.makedirs(output_dir, exist_ok=True)

    # Get the run
    run = mlflow.get_run(run_id)

    # List all artifacts in the run
    artifacts = mlflow.artifacts.list_artifacts(run_id=run_id)

    downloaded_files = {}

    for artifact in artifacts:
        if artifact.path.endswith(".csv"):
            # Download the artifact to a temporary location
            temp_path = mlflow.artifacts.download_artifacts(
                run_id=run_id,
                artifact_path=artifact.path,
                dst_path=output_dir,
            )

            # Determine the new filename based on the artifact type
            if "trajectories" in artifact.path.lower():
                new_filename = f"trajectories_{run_id}.csv"
            else:
                new_filename = f"npvs_{run_id}.csv"

            # Create the new file path
            new_path = os.path.join(output_dir, new_filename)

            # Move the file to its new name
            shutil.move(temp_path, new_path)

            downloaded_files[artifact.path] = new_path
            print(f"Downloaded and renamed {artifact.path} to {new_filename}")

    return downloaded_files


def display_runs(runs: List[Dict], title: str = ""):
    """
    Display the runs and their parameters in a readable format.
    """
    if title:
        print(f"\n{title}")
    print(f"Found {len(runs)} successful runs:\n")

    for i, run in enumerate(runs, 1):
        print(f"Run {i}:")
        print(f"  Run ID: {run['run_id']}")
        print(f"  Status: {run['status']}")
        print(f"  Start Time: {run['start_time']}")
        print(f"  End Time: {run['end_time']}")
        print("\n  Parameters:")
        for key, value in run["parameters"].items():
            print(f"    {key}: {value}")

        print("\n  Metrics:")
        for key, value in run["metrics"].items():
            print(f"    {key}: {value}")
        print("\n" + "-" * 80 + "\n")


def main(output_path: str):
    """
    Main function to fetch and download MLflow runs and artifacts.

    Args:
        output_path (str): Base directory for output files. Will create:
            - {output_path}/artifacts/ for downloaded artifacts
            - {output_path}/successful_runs.json for run information
    """
    try:
        # Create output directory structure
        output_dir = Path(output_path)
        try:
            output_dir.mkdir(parents=True, exist_ok=True)
            print(f"Created output directory: {output_dir.absolute()}")
        except Exception as e:
            raise Exception(f"Failed to create output directory {output_dir}: {str(e)}")

        artifacts_dir = output_dir / "artifacts"
        try:
            artifacts_dir.mkdir(parents=True, exist_ok=True)
            print(f"Created artifacts directory: {artifacts_dir.absolute()}")
        except Exception as e:
            raise Exception(
                f"Failed to create artifacts directory {artifacts_dir}: {str(e)}"
            )

        # Fetch successful runs
        successful_runs = fetch_successful_runs()

        # Display all runs
        display_runs(successful_runs, "All Successful Runs")

        # Filter and display runs for specific target scenario
        target_scenario = "AR6_IMAGE 3.2_SSP1_SPA1_19I_RE_LB"
        filtered_runs = filter_runs_by_target_scenario(successful_runs, target_scenario)
        display_runs(filtered_runs, f"Runs with target scenario: {target_scenario}")

        # Download artifacts for each filtered run
        print("\nDownloading artifacts for filtered runs...")
        for run in filtered_runs:
            run_id = run["run_id"]
            scenario_geo = run["parameters"].get("scenario_geography", "unknown")
            print(
                f"\nDownloading artifacts for run {run_id} (Geography: {scenario_geo})"
            )
            downloaded_files = download_run_artifacts(run_id, str(artifacts_dir))

            # Print summary of downloaded files
            print(f"Downloaded {len(downloaded_files)} files:")
            for artifact_path, local_path in downloaded_files.items():
                print(f"  - {artifact_path} -> {local_path}")

        # Save run information to JSON
        json_path = output_dir / "successful_runs.json"
        with open(json_path, "w") as f:
            json.dump(
                {"all_runs": successful_runs, "filtered_runs": filtered_runs},
                f,
                indent=2,
                default=str,
            )
        print(f"\nSaved detailed run information to '{json_path}'")

    except Exception as e:
        print(f"Error: {str(e)}")


if __name__ == "__main__":
    output_path = "workspace/mlflow_results"
    main(output_path)
